### ISSUE TITLE
Cannot run `ensurepip`

### ISSUE THREAD

**Issue by tekknolagi** [tekknolagi/skybison#2]
Repro:

1. Build skybison
2. Run `./build/bin/python -m ensurepip`

See formatting failure:

```
Traceback (most recent call last):
  File "third-party/cpython/Lib/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "third-party/cpython/Lib/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "library/builtins.py", line 3717, in exec
    _exec(code, mod, locals)
  File "/home/max/Documents/code/skybison/build/lib/skybison3.8/ensurepip/__main__.py", line 5, in <module>
    sys.exit(ensurepip._main())
  File "/home/max/Documents/code/skybison/build/lib/skybison3.8/ensurepip/__init__.py", line 210, in _main
    return _bootstrap(
  File "/home/max/Documents/code/skybison/build/lib/skybison3.8/ensurepip/__init__.py", line 129, in _bootstrap
    return _run_pip(args + [p[0] for p in _PROJECTS], additional_paths)
  File "/home/max/Documents/code/skybison/build/lib/skybison3.8/ensurepip/__init__.py", line 129, in _bootstrap
    return _run_pip(args + [p[0] for p in _PROJECTS], additional_paths)
  File "/home/max/Documents/code/skybison/build/lib/skybison3.8/ensurepip/__init__.py", line 36, in _run_pip
    sys.argv[:] = backup_argv
  File "/home/max/Documents/code/skybison/build/lib/skybison3.8/ensurepip/__init__.py", line 34, in _run_pip
    return exc.code
  File "/home/max/Documents/code/skybison/build/lib/skybison3.8/ensurepip/__init__.py", line 32, in _run_pip
    runpy.run_module("pip", run_name="__main__", alter_sys=True)
  File "third-party/cpython/Lib/runpy.py", line 207, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "third-party/cpython/Lib/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "third-party/cpython/Lib/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "third-party/cpython/Lib/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "third-party/cpython/Lib/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "library/builtins.py", line 3717, in exec
    _exec(code, mod, locals)
  File "/tmp/tmpqkm1sf1c/pip-20.1.1-py2.py3-none-any.whl/pip/__main__.py", line 26, in <module>
  File "/tmp/tmpqkm1sf1c/pip-20.1.1-py2.py3-none-any.whl/pip/_internal/cli/main.py", line 64, in main
  File "/tmp/tmpqkm1sf1c/pip-20.1.1-py2.py3-none-any.whl/pip/_internal/cli/main.py", line 60, in main
  File "/tmp/tmpqkm1sf1c/pip-20.1.1-py2.py3-none-any.whl/pip/_internal/cli/main_parser.py", line 61, in parse_command
  File "/tmp/tmpqkm1sf1c/pip-20.1.1-py2.py3-none-any.whl/pip/_internal/cli/main_parser.py", line 51, in create_main_parser
  File "library/builtins.py", line 6488, in format
    return _formatter.format(self, *args, **kwargs)
  File "/home/max/Documents/code/skybison/build/lib/skybison3.8/string.py", line 163, in format
    return self.vformat(format_string, args, kwargs)
  File "/home/max/Documents/code/skybison/build/lib/skybison3.8/string.py", line 167, in vformat
    result, _ = self._vformat(format_string, args, kwargs, used_args, 2)
  File "/home/max/Documents/code/skybison/build/lib/skybison3.8/string.py", line 208, in _vformat
    obj, arg_used = self.get_field(field_name, args, kwargs)
  File "/home/max/Documents/code/skybison/build/lib/skybison3.8/string.py", line 273, in get_field
    obj = self.get_value(first, args, kwargs)
  File "/home/max/Documents/code/skybison/build/lib/skybison3.8/string.py", line 230, in get_value
    return kwargs[key]
KeyError: 'name'
```

While `formatting '{name:27} {command_info.summary}' and looking for 'name'`

I am not sure what this colon syntax is.

**Comment by tekknolagi** [comment]
If I disable the comprehension renamer in `library/_compiler`, we can progress and completely install setuptools.

```patch
diff --git a/library/_compiler.py b/library/_compiler.py
index afaef834..59c2aec7 100644
--- a/library/_compiler.py
+++ b/library/_compiler.py
@@ -211,67 +211,6 @@ class PyroFlowGraph(PyFlowGraph38):
     opcode = opcodepyro.opcode
 
 
-class ComprehensionRenamer(ASTVisitor):
-    def __init__(self, scope):
-        super().__init__()
-        # We need a prefix that is unique per-scope for each renaming round.
-        index = getattr(scope, "last_comprehension_rename_index", -1) + 1
-        scope.last_comprehension_rename_index = index
-        self.prefix = f"_gen{str(index) if index > 0 else ''}$"
-        self.new_names = {}
-        self.is_target = False
-
-    def visitName(self, node):
-        if self.is_target and isinstance(node.ctx, (ast.Store, ast.Del)):
-            name = node.id
-            new_name = self.prefix + name
-            self.new_names[name] = new_name
-            node.id = new_name
-        else:
-            new_name = self.new_names.get(node.id)
-            if new_name is not None:
-                node.id = new_name
-
-    def visitarg(self, node):
-        new_name = self.new_names.get(node.arg)
-        if new_name is not None:
-            node.arg = new_name
-
-
-class PyroSymbolVisitor(SymbolVisitor):
-    def visitDictCompListCompSetComp(self, node, scope):
-        # Check for unexpected assignments.
-        scope.comp_iter_expr += 1
-        self.visit(node.generators[0].iter, scope)
-        scope.comp_iter_expr -= 1
-
-        renamer = ComprehensionRenamer(scope)
-        is_outer = True
-        for gen in node.generators:
-            renamer.visit(gen.iter)
-            renamer.is_target = True
-            renamer.visit(gen.target)
-            renamer.is_target = False
-            for if_node in gen.ifs:
-                renamer.visit(if_node)
-
-            self.visitcomprehension(gen, scope, is_outer)
-            is_outer = False
-
-        if isinstance(node, ast.DictComp):
-            renamer.visit(node.value)
-            renamer.visit(node.key)
-            self.visit(node.value, scope)
-            self.visit(node.key, scope)
-        else:
-            renamer.visit(node.elt)
-            self.visit(node.elt, scope)
-
-    visitDictComp = visitDictCompListCompSetComp
-    visitListComp = visitDictCompListCompSetComp
-    visitSetComp = visitDictCompListCompSetComp
-
-
 class PyroCodeGenerator(Python38CodeGenerator):
     flow_graph = PyroFlowGraph
 
@@ -288,7 +227,7 @@ class PyroCodeGenerator(Python38CodeGenerator):
     ):
         if ast_optimizer_enabled:
             tree = cls.optimize_tree(optimize, tree)
-        s = PyroSymbolVisitor()
+        s = SymbolVisitor()
         walk(tree, s)
 
         graph = cls.flow_graph(
@@ -310,20 +249,6 @@ class PyroCodeGenerator(Python38CodeGenerator):
         else:
             self.emit("COMPARE_OP", self._cmp_opcode[type(op)])
 
-    def visitListComp(self, node):
-        self.emit("BUILD_LIST")
-        self.compile_comprehension_body(node.generators, 0, node.elt, None, type(node))
-
-    def visitSetComp(self, node):
-        self.emit("BUILD_SET")
-        self.compile_comprehension_body(node.generators, 0, node.elt, None, type(node))
-
-    def visitDictComp(self, node):
-        self.emit("BUILD_MAP")
-        self.compile_comprehension_body(
-            node.generators, 0, node.key, node.value, type(node)
-        )
-
 
 def compile(source, filename, mode, flags, dont_inherit, optimize):
     return compiler_compile(
```

This is not ideal because the comprehension renamer is a big performance win.

**Comment by tekknolagi** [comment]
This is due to the dict comprehension in https://github.com/pypa/pip/blob/8356bc5161e0a9b3054f0e04e12d18feae4c2b46/src/pip/_internal/cli/main_parser.py#L50

**Comment by tekknolagi** [comment]
Small, fast repro:

```python
def foo():
    commands_dict = {"NAME": "INFO"}
    return [
        '{name} {command_info}'.format(**locals())
        for name, command_info in commands_dict.items()
    ]

foo()
```

**Comment by tekknolagi** [comment]
It might be possible to disable the comprehension renamer in the presence of a call to `locals`, but I am not 100% how to do this...

**Comment by tekknolagi** [comment]
Disabled it wholesale for now.

**Comment by tekknolagi** [comment]
Fixed in #18 (266fc9fd8e1c40d479e2f39faea7a6fc68267803)
